File: packages/platform/src/crypto/deployment-scope.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { createHmac, randomBytes } from "node:crypto";
import fs from "node:fs";

const DEPLOYMENT_SCOPE_VERSION = "v1";
// Keep the grant window short enough to limit replay after a worker
// compromise, while allowing the broker's 30-minute build operations and
// deployment retries to complete without minting a second grant.
const DEPLOYMENT_SCOPE_TTL_MS = 2 * 60 * 60 * 1_000;
const MINIMUM_SCOPE_SECRET_BYTES = 32;
const DEPLOYMENT_SCOPE_IDENTIFIER_PATTERN =
  /^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/;
const deploymentScopeStorage = new AsyncLocalStorage<{
  token?: string;
}>();

export type DeploymentScopeClaims = {
  resourceId: string;
  deploymentId: string;
  serverId: string;
  issuedAt: number;
  expiresAt: number;
  nonce: string;
};

function readScopeSecret(): Buffer | undefined {
  const secretFile =
    process.env.UPSTAND_DOCKER_BROKER_SCOPE_SECRET_FILE?.trim();
  const value = secretFile
    ? fs.readFileSync(secretFile)
    : process.env.UPSTAND_DOCKER_BROKER_SCOPE_SECRET;
  if (!value) return undefined;
  const secret = Buffer.isBuffer(value)
    ? value
    : Buffer.from(value.trim(), "utf8");
  if (secret.length < MINIMUM_SCOPE_SECRET_BYTES) {
    throw new Error(
      "UPSTAND_DOCKER_BROKER_SCOPE_SECRET must contain at least 32 bytes",
    );
  }
  return secret;
}

/**
 * Creates a short-lived grant for one deployment and resource. The private
 * secret is intentionally read only by the control-plane queueing processes;
 * deployment workers receive the resulting grant through the queue payload.
 */
export function createDeploymentScopeToken(input: {
  resourceId: string;
  deploymentId: string;
  serverId: string;
  now?: number;
}): string | undefined {
  for (const [name, value] of Object.entries({
    resourceId: input.resourceId,
    deploymentId: input.deploymentId,
    serverId: input.serverId,
  })) {
    if (!DEPLOYMENT_SCOPE_IDENTIFIER_PATTERN.test(value)) {
      throw new Error(`Deployment scope ${name} is invalid`);
    }
  }
  const secret = readScopeSecret();
  if (!secret) return undefined;
  const issuedAt = input.now ?? Date.now();
  const claims: DeploymentScopeClaims = {
    resourceId: input.resourceId,
    deploymentId: input.deploymentId,
    serverId: input.serverId,
    issuedAt,
    expiresAt: issuedAt + DEPLOYMENT_SCOPE_TTL_MS,
    nonce: randomBytes(16).toString("base64url"),
  };
  const payload = Buffer.from(JSON.stringify(claims), "utf8").toString(
    "base64url",
  );
  const signature = createHmac("sha256", secret)
    .update(`${DEPLOYMENT_SCOPE_VERSION}.${payload}`)
    .digest("base64url");
  return `${DEPLOYMENT_SCOPE_VERSION}.${payload}.${signature}`;
}

/** Returns the signed scope attached to the current asynchronous deployment. */
export function readDeploymentScopeToken(): string | undefined {
  const scoped = deploymentScopeStorage.getStore();
  if (scoped) return scoped.token;
  return process.env.UPSTAND_DEPLOYMENT_SCOPE_TOKEN?.trim() || undefined;
}

/**
 * Returns the signed grant and its routing claims for broker transports.
 * Claims are only used to mirror the grant binding into request headers; the
 * broker remains the authority that verifies the signature and lifetime.
 */
export function readDeploymentScopeHeaders(): Record<string, string> {
  const token = readDeploymentScopeToken();
  if (!token) return {};

  const parts = token.split(".");
  const payload = parts[1];
  if (
    parts.length !== 3 ||
    parts[0] !== DEPLOYMENT_SCOPE_VERSION ||
    payload === undefined
  ) {
    return { "X-Upstand-Docker-Scope": token };
  }

  try {
    const claims = JSON.parse(
      Buffer.from(payload, "base64url").toString("utf8"),
    ) as Partial<DeploymentScopeClaims>;
    const headers: Record<string, string> = {
      "X-Upstand-Docker-Scope": token,
    };
    if (typeof claims.deploymentId === "string" && claims.deploymentId) {
      headers["X-Upstand-Deployment-ID"] = claims.deploymentId;
    }
    if (typeof claims.serverId === "string" && claims.serverId) {
      headers["X-Upstand-Server-ID"] = claims.serverId;
    }
    return headers;
  } catch {
    return { "X-Upstand-Docker-Scope": token };
  }
}

/** Runs a deployment while making its grant available to broker transports. */
export async function withDeploymentScopeToken<T>(
  token: string | undefined,
  operation: () => Promise<T>,
): Promise<T> {
  return deploymentScopeStorage.run({ token }, operation);
}

File: apps/docker-broker/deployment_scope.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

const (
	deploymentScopeHeader = "X-Upstand-Docker-Scope"
	deploymentIDHeader    = "X-Upstand-Deployment-ID"
	serverIDHeader        = "X-Upstand-Server-ID"
	minimumScopeSecretLen = 32
	// Keep signed worker grants short-lived to reduce replay exposure after a
	// worker compromise while covering the broker's bounded build operations.
	maximumScopeLifetime = 2 * time.Hour
	maximumClockSkew     = 5 * time.Minute
)

var deploymentScopeIdentifierPattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$`)

type deploymentScopeClaims struct {
	ResourceID   string `json:"resourceId"`
	DeploymentID string `json:"deploymentId"`
	ServerID     string `json:"serverId"`
	IssuedAt     int64  `json:"issuedAt"`
	ExpiresAt    int64  `json:"expiresAt"`
	Nonce        string `json:"nonce"`
}

func loadDockerScopeSecret() ([]byte, error) {
	file := strings.TrimSpace(os.Getenv("UPSTAND_DOCKER_BROKER_SCOPE_SECRET_FILE"))
	var value []byte
	var err error
	if file != "" {
		value, err = os.ReadFile(file)
		if err != nil {
			return nil, fmt.Errorf("read Docker broker scope secret: %w", err)
		}
	} else {
		value = []byte(strings.TrimSpace(os.Getenv("UPSTAND_DOCKER_BROKER_SCOPE_SECRET")))
	}
	value = []byte(strings.TrimSpace(string(value)))
	if len(value) == 0 {
		if brokerRequiresProductionIdentity() {
			return nil, errors.New("Docker broker scope signing secret is required in production")
		}
		return nil, nil
	}
	if len(value) < minimumScopeSecretLen {
		return nil, errors.New("Docker broker scope signing secret must contain at least 32 bytes")
	}
	return value, nil
}

// authorizeDeploymentWorkerScopeToken verifies that the deployment worker was
// granted the resource scope by a queueing control-plane process. The worker
// can forward a grant, but it cannot mint a grant for another resource because
// it never receives the signing secret.
func authorizeDeploymentWorkerScopeToken(caller string, r *http.Request, body, secret []byte) error {
	if caller != "deployment-worker" || !brokerRequiresProductionIdentity() {
		return nil
	}
	if len(secret) < minimumScopeSecretLen {
		return errors.New("deployment-worker scope signing is not configured")
	}
	token := strings.TrimSpace(r.Header.Get(deploymentScopeHeader))
	if token == "" {
		return errors.New("deployment-worker request requires a signed deployment scope")
	}
	claims, suppliedSignature, signedValue, err := parseDeploymentScopeToken(token)
	if err != nil {
		return err
	}
	expectedMAC := hmac.New(sha256.New, secret)
	_, _ = expectedMAC.Write([]byte(signedValue))
	expectedSignature := expectedMAC.Sum(nil)
	decodedSignature, err := base64.RawURLEncoding.DecodeString(suppliedSignature)
	if err != nil || !hmac.Equal(expectedSignature, decodedSignature) {
		return errors.New("deployment-worker deployment scope signature is invalid")
	}

	now := time.Now()
	issuedAt := time.UnixMilli(claims.IssuedAt)
	expiresAt := time.UnixMilli(claims.ExpiresAt)
	if claims.IssuedAt <= 0 || claims.ExpiresAt <= claims.IssuedAt ||
		expiresAt.Before(now) || issuedAt.After(now.Add(maximumClockSkew)) ||
		expiresAt.Sub(issuedAt) > maximumScopeLifetime {
		return errors.New("deployment-worker deployment scope is expired or has invalid lifetime")
	}
	if !resourceIDPattern.MatchString(claims.ResourceID) ||
		!deploymentScopeIdentifierPattern.MatchString(claims.DeploymentID) ||
		!deploymentScopeIdentifierPattern.MatchString(claims.ServerID) ||
		!deploymentScopeIdentifierPattern.MatchString(claims.Nonce) {
		return errors.New("deployment-worker deployment scope claims are invalid")
	}

	resourceID := strings.TrimSpace(r.Header.Get("X-Upstand-Resource-ID"))
	if resourceID == "" {
		resourceID = resourceIDFromBody(body)
	}
	if resourceID != "" && resourceID != claims.ResourceID {
		return errors.New("deployment-worker resource scope does not match the signed deployment grant")
	}
	if strings.TrimSpace(r.Header.Get(deploymentIDHeader)) != claims.DeploymentID {
		return errors.New("deployment-worker deployment scope does not match the signed deployment grant")
	}
	if strings.TrimSpace(r.Header.Get(serverIDHeader)) != claims.ServerID {
		return errors.New("deployment-worker server scope does not match the signed deployment grant")
	}
	return nil
}

func parseDeploymentScopeToken(token string) (deploymentScopeClaims, string, string, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 || parts[0] != "v1" || parts[1] == "" || parts[2] == "" {
		return deploymentScopeClaims{}, "", "", errors.New("deployment-worker deployment scope format is invalid")
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return deploymentScopeClaims{}, "", "", errors.New("deployment-worker deployment scope payload is invalid")
	}
	var claims deploymentScopeClaims
	if err := json.Unmarshal(payload, &claims); err != nil {
		return deploymentScopeClaims{}, "", "", errors.New("deployment-worker deployment scope claims are invalid")
	}
	return claims, parts[2], parts[0] + "." + parts[1], nil
}

func resourceIDFromBody(body []byte) string {
	if len(strings.TrimSpace(string(body))) == 0 {
		return ""
	}
	var payload struct {
		ResourceID string `json:"resource_id"`
	}
	if json.Unmarshal(body, &payload) != nil {
		return ""
	}
	return strings.TrimSpace(payload.ResourceID)
}
